bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LedgerDirsMonitor.java

```diff
@@ -0,0 +1,191 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package org.apache.bookkeeper.bookie;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+
+import org.apache.bookkeeper.bookie.LedgerDirsManager.LedgerDirsListener;
+import org.apache.bookkeeper.bookie.LedgerDirsManager.NoWritableLedgerDirException;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.util.DiskChecker;
+import org.apache.bookkeeper.util.DiskChecker.DiskErrorException;
+import org.apache.bookkeeper.util.DiskChecker.DiskOutOfSpaceException;
+import org.apache.bookkeeper.util.DiskChecker.DiskWarnThresholdException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Thread to monitor the disk space periodically.
+ */
+class LedgerDirsMonitor extends BookieThread {
+    private final static Logger LOG = LoggerFactory.getLogger(LedgerDirsMonitor.class);
+    
+    private final int interval;
+    private final ServerConfiguration conf;
+    private final ConcurrentMap<File, Float> diskUsages;
+    private final DiskChecker diskChecker;
+    private final LedgerDirsManager ldm;
+
+    public LedgerDirsMonitor(final ServerConfiguration conf, 
+            final DiskChecker diskChecker,
+            final LedgerDirsManager ldm) {
+        super("LedgerDirsMonitorThread");
+        this.interval = conf.getDiskCheckInterval();
+        this.conf = conf;
+        this.diskChecker = diskChecker;
+        this.diskUsages = ldm.getDiskUsages();
+        this.ldm = ldm;
+    }
+
+    @Override
+    public void run() {
+        while (true) {
+            try {
+                List<File> writableDirs = ldm.getWritableLedgerDirs();
+                // Check all writable dirs disk space usage.
+                for (File dir : writableDirs) {
+                    try {
+                        diskUsages.put(dir, diskChecker.checkDir(dir));
```

Comment:
I agree that circular dependencies are bad, but so is having one class reach into the guts of another class to modify its internal state.  I do not see any reduction in coupling between the classes.  LedgerDirsManager is broken without a LedgerDirsMonitor also running (most of the callbacks just will not happen without it).  LedgerDirsMonitor cannot be instantiated without a LedgerDirsManager for it to go reach into.  All I see is that a single file was broken into two files and some APIs were made public that were not before.

To me the original LedgerDirsMonitor class does not really feel like a strict circular dependency.  It feels like an implementation detail of LedgerDirsManager that is hidden (private).

If you really want them separate then make a separate LedgerDirsMonitorCallback class.  Have the Monitor then call into instances of that callback that are registered with it.  Then have the LedgerDirsManager take a Monitor class in its constructor and register itself with it. The Manager can listen to the Monitor callbacks then update its own state internally and make its own callbacks.  This way you can never instantiate a LedgerDirsManager that will not function properly, and LedgerDirsMonitor has no knowledge of the LedgerDirsManager.